### ISSUE TITLE
Improve the filtering experience: add new operators and change some behaviors

### ISSUE THREAD

**Issue by curquiza** [meilisearch/meilisearch#2580]
The filtering is powerful and wildly feature of Meilisearch. By gathering some user feedback, we want to improve this experience.

This central issue gathers the different issues opened about the filtering
- https://github.com/meilisearch/meilisearch/issues/2486
- https://github.com/meilisearch/meilisearch/issues/2479
- https://github.com/meilisearch/meilisearch/issues/2484
- https://github.com/meilisearch/meilisearch/issues/2516
- https://github.com/meilisearch/meilisearch/issues/2679

Here are the related product discussions
- https://github.com/meilisearch/product/discussions/479
- https://github.com/meilisearch/product/discussions/274
- https://github.com/meilisearch/product/discussions/407

## Process

💫 This issue will be completed with links and all the necessary information all along the feature creation process

- [X] Create a prototype by providing a docker image

A docker tag is available: 

```bash
docker run -it --rm \
    -p 7700:7700 \
    getmeili/meilisearch:v0.29.0-filter.beta.0
```

Or you can compile the source code on the `filter/field-in` branch on this repo

All the information about the new addition is detailed [here](https://github.com/meilisearch/product/discussions/407#discussioncomment-3084053)

- [ ] Apply changes according to feedback
- [X] Create the corresponding specification and link it to this issue -> https://github.com/meilisearch/specifications/pull/163

⚠️ if the discovery phase (prototype creation + user feedback) is not satisfying, we will postpone this feature addition

**Comment by bidoubiwa** [comment]
Hey @curquiza!  Thanks for the ping :) How can we help during the discovery phase ? We are currently working on finding a process to create releases with specific features that we want to let users try during the discovery phase. 
Would this feature be something you would like to see implemented in a SDK and communicated to users so that they can try it more simply?

**Comment by loiclec** [comment]
Note that after these changes are merged, the [filter depth limitation]( https://docs.meilisearch.com/learn/advanced/known_limitations.html#maximum-filter-depth) will be eliminated, and thus the documentation will need to be updated :)  @curquiza

**Comment by curquiza** [comment]
Is the limitation completely removed? Aren't the parenthesis (`[]`) increasing the filter depth?

**Comment by loiclec** [comment]
Good point! I had to double-check.

1. if the filters are given as arrays instead of strings, the limitations are also removed.
2. but there remains a potential stack overflow for the filter: `(((((...(((((((x = 1)))))))...)))))`. The stack overflow is not caught at the moment though, so it always crashes meilisearch instead of returning an error.

**Comment by curquiza** [comment]
FYI, the corresponding spec has been opened by @loiclec 🎉  -> https://github.com/meilisearch/specifications/pull/163

**Comment by curquiza** [comment]
Better late than never, a prototype is available, see how to use this issue description

**Comment by mech** [comment]
If I use the docker for pagination like **getmeili/meilisearch:v0.29.0-pagination.beta.3**, does it also contain this filtering prototype. Or I need to use 2 different dockers to experiment?

**Comment by curquiza** [comment]
Hello @mech

Sorry, only this docker image put in the issue description work for filter: `getmeili/meilisearch:v0.29.0-filter.beta.0`
However, the first RC of v0.29.0 is out next Monday, you will be able to test the full v0.29.0 release. But it does not contain the pagination feature we delayed due to a [lack of feedback on it](https://github.com/meilisearch/meilisearch/discussions/2635); we want to be sure before releasing it.

**Comment by curquiza** [comment]
Hello everyone watching this issue!

We have just released [v0.29.0rc1](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1), which is a release candidate of v0.29.0 🔥 
It means you can test the just-introduced `EXISTS` and `IN` filtering operators! [Binaries](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1) are attached to the release, or you can you the docker image:
```bash
docker run -it --rm \
    -p 7700:7700 \
    getmeili/meilisearch:v0.29.0rc0
```

Let us know about any bugs or feedback! 😄 It would be really helpful.

FYI, the official v0.29.0 release will be available on 3rd October.

**Comment by mech** [comment]
If I am not wrong, this RC1 release is only having the filtering prototype right, without the pagination prototype.
